### ISSUE TITLE
[HOLD for payment 2023-05-16] [$1000] iOS - Chat - Can't download webp format image after it was upload to chat

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#16599]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/16031

## Action Performed:
1. Open the app
2. Login with any account
3. Go to any chat
4. Put on + button -> Add attachment -> send a webp format image as a file
5. Open uploaded attachment -> download it

## Expected Result:
An attachment is download

## Actual Result:
Can't download attachment, an error appears

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.90.4

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/228081066-5462eea5-9a44-4d7c-98ec-4d659b69488a.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ce67f235c93f2d4c</li>
        <li>Upwork Job ID: 1643450952704184320</li>
        <li>Last Price Increase: 2023-04-19</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @JmillsExpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by JmillsExpensify** [comment]
@youssef-lr I kind of wonder if we should put this issue on hold and handle it as part of supporting more file types, basically more formally than we do today? At the moment, I'm not really sure that this is technically a bug.

**Comment by MelvinBot** [comment]
@JmillsExpensify Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@JmillsExpensify Still overdue 6 days?! Let's take care of this!

**Comment by JmillsExpensify** [comment]
Ok coming back to this. I wasn't able to successfully reproduce this. I would normally say we shouldn't bother, but we do officially support this file format, so I'm going to open this up for proposals.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01ce67f235c93f2d4c

**Comment by MelvinBot** [comment]
Current assignee @JmillsExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @sobitneupane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @marcochavezf (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by akinwale** [comment]
I was unable to reproduce this on the latest main branch.

**Comment by akinwale** [comment]
> It seems it related to some IOS devices where it's not support webp

If this were the case, the image wouldn't have even displayed in the first place. webp images are not supported on iOS prior to v14.

**Comment by marcochavezf** [comment]
@kbecciv could you try to reproduce this again? Seems this is no longer an issue (maybe was fixed recently?)

**Comment by kbecciv** [comment]
@marcochavezf Let me check with QA team, will post you results shortly

**Comment by hellohublot** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

iOS - Chat - Can't download webp format image after it was upload to chat

### What is the root cause of that problem?

iOS cannot save webp photos

### What changes do you think we should make in order to solve the problem?

iOS 14+ supports decoding webp images, but the support is not very good

Solution A
- We do not perform format conversion, we can save it in the file app, but we may need to give the user a little hint to find it in the file app
https://github.com/Expensify/App/blob/3abe7b2f9e38dba42edf406ca1b1e19087f5599d/src/libs/fileDownload/index.ios.js#L74-L75
```javascript
if (/\.webp$/.test(attachmentName)) {
    fileType = CONST.ATTACHMENT_FILE_TYPE.FILE
}
...
FileUtils.showSuccessAlert('you can find it in File app Expensify floder')
```
---
Solution B.
- We can use cloudfront for format conversion, and the backend can add a format parameter to the image url and forward it to cloudfront
https://github.com/Expensify/App/blob/3abe7b2f9e38dba42edf406ca1b1e19087f5599d/src/libs/fileDownload/index.ios.js#L74-L75
```javascript
if (/\.webp$/.test(attachmentName)) {
    fileUrl += '&format=jpeg'
}
```
---
Solution C.
- We can perform `format conversion`, if `SDWebImage+WebP` can be imported, we use it to `decode`, we use it to `decode`, then we convert to `jpeg`, and then we save
https://github.com/react-native-cameraroll/react-native-cameraroll/blob/d12368ba7326c696a4fb25cd878c77efcb822f8c/ios/RNCCameraRoll.mm#L166
```objective-c
} else if ([[inputURI.pathExtension lowercaseString] isEqualToString:@"webp"]) {
    NSData *data = [NSData dataWithContentsOfURL:inputURI];
    UIImage *image;
    #if __has_include(<SDWebImageWebPCoder/UIImage+WebP.h>)
	image = [UIImage sd_imageWithWebPData:data];
    #else
	image = [UIImage imageWithData:data];
    #endif

    image = [UIImage imageWithData:UIImageJPEGRepresentation(image, 1)];

    assetRequest = [PHAssetChangeRequest creationRequestForAssetFromImage:image];
}
```


### What alternative solutions did you explore? (Optional)
Not Yet

**Comment by MelvinBot** [comment]
@JmillsExpensify @marcochavezf @sobitneupane this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
@JmillsExpensify @marcochavezf @sobitneupane this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by sobitneupane** [comment]
> Let me check with QA team, will post you results shortly

@kbecciv Was the QA team able to reproduce the issue?

**Comment by kbecciv** [comment]
Issue is still reproductible on build 1.2.97.2


https://user-images.githubusercontent.com/93399543/230937730-88f9242d-a1e3-4113-aee3-6ef74b8a69e2.MP4

**Comment by sobitneupane** [comment]
Gentle bump @marcochavezf https://github.com/Expensify/App/issues/16599#issuecomment-1502966878

**Comment by marcochavezf** [comment]
> Issue is still reproductible on build 1.2.97.2

Oh interesting, @kbecciv do you know which version of iOS or iPhone was used to reproduce the issue?

**Comment by marcochavezf** [comment]
I try to reproduce it but the webp images load fine for me. I think we should close this issue if the bug only happens only in old iPhone devices.

**Comment by MelvinBot** [comment]
@JmillsExpensify @marcochavezf @sobitneupane this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by JmillsExpensify** [comment]
@marcochavezf Are you sure it's only old iPhones? This bug is not about uploading webp photos but downloading them. For that part, I am able to reproduce.

**Comment by marcochavezf** [comment]
> @marcochavezf Are you sure it's only old iPhones? This bug is not about uploading webp photos but downloading them. For that part, I am able to reproduce.

The [Apple documentation](https://developer.apple.com/documentation/uniformtypeidentifiers/uttype/3551599-webp) only mentions iOS 14+ for webp support, and some of us are unable to reproduce it. That's why I thought it happened on old iPhones. @kbecciv could you confirm if the issue is [reproducible](https://github.com/Expensify/App/issues/16599#issuecomment-1501972914) on iOS 14+?

**Comment by marcochavezf** [comment]
> Should we also do the conversion? Should we convert it across all platforms or just for IOS? To which format should it be converted?

If the issue is only happening on iOS, I think we should fix it there only.

**Comment by sobitneupane** [comment]
@hellohublot Will you please update your proposal based on https://github.com/Expensify/App/issues/16599#issuecomment-1517123274. We should allow users to download webp format in other devices and should perform conversion only in IOS.

**Comment by JmillsExpensify** [comment]
@marcochavezf Ah ok cool, thanks for the clarification. I agree with only fixing this on iOS only.

**Comment by hellohublot** [comment]
@sobitneupane Thanks

I have removed my [Solution A](https://github.com/Expensify/App/issues/16599#issuecomment-1501086016), but I still keep Solution B available
https://github.com/react-native-cameraroll/react-native-cameraroll/pull/489
I have submitted a PR for Solution C, if my PR can't be accepted, So solution B is still available

**Comment by MelvinBot** [comment]
@JmillsExpensify @marcochavezf @sobitneupane this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @sobitneupane is eligible for the Internal assigner, not assigning anyone new.

**Comment by sobitneupane** [comment]
Looks like [associated PR](https://github.com/react-native-cameraroll/react-native-cameraroll/pull/489) on https://github.com/react-native-cameraroll/react-native-cameraroll repo is under review.

**Comment by hellohublot** [comment]
@sobitneupane Thanks, The upstream pr has been merged and published, you can try it now `"@react-native-camera-roll/camera-roll": "5.4.0",`

**Comment by marcochavezf** [comment]
Thanks for the upstream @hellohublot! Could you create a PR to update the camera-roll npm package?

**Comment by MelvinBot** [comment]
Current assignee @JmillsExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @sobitneupane is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @marcochavezf is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
📣 @hellohublot You have been assigned to this job by @marcochavezf!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

Issue not related to our App. Issue was existent on https://github.com/react-native-cameraroll/react-native-cameraroll and was solved by https://github.com/react-native-cameraroll/react-native-cameraroll/pull/489

**Comment by JmillsExpensify** [comment]
Alright, circling back on this issue. It looks like our payouts should be:

 - Contributor: @hellohublot $1,500
 - Contributor: @sobitneupane $1,500
 
I've already invited Sobit. @hellohublot mind sharing your Upwork profile so I can hire you as well? 

P.S. @sobitneupane I think [this approach](https://github.com/Expensify/App/issues/16599#issuecomment-1551458109) sounds reasonable.

**Comment by hellohublot** [comment]
@JmillsExpensify Thanks. https://www.upwork.com/freelancers/~01509ca2c74bfc7073

**Comment by JmillsExpensify** [comment]
Apologies for the delay guys. I went OOO and I'm just coming back. I'm closing the loop on the payments now.

**Comment by JmillsExpensify** [comment]
@sobitneupane has been paid out and I just invited @hellohublot. As soon as you're able to answer the invite, I'll issue contract, payment and bonus. Thanks!

**Comment by JmillsExpensify** [comment]
Thanks for the quick response. Everyone should have been paid out now and now regression tests are outstanding. I'm closing this one out!
